aliasmgr/aliasmgr.go

```diff
@@ -258,6 +269,10 @@ func (m *Manager) AddLocalAlias(alias, baseScid lnwire.ShortChannelID,
 		byteOrder.PutUint64(baseBytes[:], baseScid.ToUint64())
 		return aliasToBaseBucket.Put(aliasBytes[:], baseBytes[:])
 	}, func() {})
+
+	// Before checking for the returned error, we'll unlock the Manager.
+	m.Unlock()
```

Comment:
I think we need to hold the lock longer to avoid access to the maps below to be corrupted.
Here's my suggestion:
```diff
diff --git a/aliasmgr/aliasmgr.go b/aliasmgr/aliasmgr.go
index 813b3f6dc..faf8e8c9d 100644
--- a/aliasmgr/aliasmgr.go
+++ b/aliasmgr/aliasmgr.go
@@ -231,7 +231,16 @@ func (m *Manager) populateMaps() error {
 func (m *Manager) AddLocalAlias(alias, baseScid lnwire.ShortChannelID,
        gossip bool) error {
 
+       // We need to lock the manager for the whole duration of this method,
+       // except for the very last part where we call the link updater. In
+       // order for us to safely use a defer _and_ still be able to manually
+       // unlock, we use a sync.Once.
        m.Lock()
+       unlockOnce := sync.Once{}
+       unlock := func() {
+               unlockOnce.Do(m.Unlock)
+       }
+       defer unlock()
 
        err := kvdb.Update(m.backend, func(tx kvdb.RwTx) error {
                // If the caller does not want to allow the alias to be used
@@ -269,10 +278,6 @@ func (m *Manager) AddLocalAlias(alias, baseScid lnwire.ShortChannelID,
                byteOrder.PutUint64(baseBytes[:], baseScid.ToUint64())
                return aliasToBaseBucket.Put(aliasBytes[:], baseBytes[:])
        }, func() {})
-
-       // Before checking for the returned error, we'll unlock the Manager.
-       m.Unlock()
-
        if err != nil {
                return err
        }
@@ -285,14 +290,14 @@ func (m *Manager) AddLocalAlias(alias, baseScid lnwire.ShortChannelID,
                m.aliasToBase[alias] = baseScid
        }
 
-       // Finally, we trigger an htlcswitch update, in order for any future
-       // htlc that references the added alias to be properly routed.
-       err = m.LinkUpdater(baseScid)
-       if err != nil {
-               return err
-       }
+       // We definitely need to unlock the Manager before calling the link
+       // updater. If we don't, we'll deadlock. We use a sync.Once to ensure
+       // that we only unlock once.
+       unlock()
 
-       return nil
+       // Finally, we trigger a htlcswitch update, in order for any future
+       // htlc that references the added alias to be properly routed.
+       return m.LinkUpdater(baseScid)
 }
 
 // GetAliases fetches the set of aliases stored under a given base SCID from
@@ -366,8 +371,16 @@ func (m *Manager) DeleteSixConfs(baseScid lnwire.ShortChannelID) error {
 func (m *Manager) DeleteLocalAlias(alias,
        baseScid lnwire.ShortChannelID) error {
 
+       // We need to lock the manager for the whole duration of this method,
+       // except for the very last part where we call the link updater. In
+       // order for us to safely use a defer _and_ still be able to manually
+       // unlock, we use a sync.Once.
        m.Lock()
-       defer m.Unlock()
+       unlockOnce := sync.Once{}
+       unlock := func() {
+               unlockOnce.Do(m.Unlock)
+       }
+       defer unlock()
 
        err := kvdb.Update(m.backend, func(tx kvdb.RwTx) error {
                aliasToBaseBucket, err := tx.CreateTopLevelBucket(aliasBucket)
@@ -412,6 +425,11 @@ func (m *Manager) DeleteLocalAlias(alias,
        // cache.
        delete(m.aliasToBase, alias)
 
+       // We definitely need to unlock the Manager before calling the link
+       // updater. If we don't, we'll deadlock. We use a sync.Once to ensure
+       // that we only unlock once.
+       unlock()
+
        return m.LinkUpdater(baseScid)
 }
```

Reply:
nice pattern, will add! :+1: